pallets/msa/src/lib.rs

```diff
@@ -1292,6 +1353,99 @@ impl<T: Config> Pallet<T> {
+		events: Vec<(T::AccountId, EventType, T::BlockNumber)>,
+	) {
+		// Lock will specifically prevent multiple offchain workers from
+		// processing the same msa events at the same time
+		let lock_status = offchain_common::lock(b"msa::", msa_id.encode().as_slice(), || {
+			for (key, event_type, block) in &events {
+				match event_type {
+					EventType::Add => {
+						let add_result =
+							process_msa_key_event::<MessageSourceId, T::AccountId, T::BlockNumber>(
+								MSAPublicKeyDataOperation::Add::<
+									MessageSourceId,
+									T::AccountId,
+									T::BlockNumber,
+								>(msa_id, key.clone(), *block),
+							);
+						if let Err(e) = add_result {
+							log_err!("Error adding key to offchain storage: {:?}", e);
+						}
+					},
+					EventType::Remove => {
+						let delete_result =
+							process_msa_key_event::<MessageSourceId, T::AccountId, T::BlockNumber>(
+								MSAPublicKeyDataOperation::Remove::<
+									MessageSourceId,
+									T::AccountId,
+									T::BlockNumber,
+								>(msa_id, key.clone(), *block),
+							);
+						if let Err(e) = delete_result {
+							log_err!("Error deleting key from offchain storage: {:?}", e);
+						}
+					},
+				}
+			}
+		});
+		if let offchain_common::LockStatus::Locked = lock_status {
+			log_err!("Unable to acquire lock, skipping event processing");
+		}
+	}
+
+	fn index_event(event: Event<T>) {
+		let block_number: u32 =
+			<frame_system::Pallet<T>>::block_number().try_into().unwrap_or_default();
+		let current_event_count: u16 = <MSAEventCount<T>>::get().saturating_add(1);
+		<MSAEventCount<T>>::put(current_event_count);
+		let event_key = [
+			BLOCK_EVENT_KEY,
+			block_number.encode().as_slice(),
+			current_event_count.encode().as_slice(),
+		]
+		.concat();
+		// set the event in offchain storage
+		set_offchain_index(event_key.encode().as_slice(), event);
+
+		let count_key = [BLOCK_EVENT_COUNT_KEY, block_number.encode().as_slice()].concat();
+		// Set the latest count of event in current block
+		set_offchain_index(count_key.encode().as_slice(), current_event_count);
+	}
+
+	fn read_events(block_number: T::BlockNumber) -> Vec<Event<T>> {
+		let current_block: u32 = block_number.try_into().unwrap_or_default();
+		let count_key = [BLOCK_EVENT_COUNT_KEY, current_block.encode().as_slice()].concat();
+		let optional_event_count = get_offchain_index::<u16>(count_key.encode().as_slice());
+		let mut events = vec![];
+		let event_count =
+			if let Some(event_count) = optional_event_count { event_count } else { return events };
+
+		for i in 1..event_count + 1 {
+			let key =
+				[BLOCK_EVENT_KEY, block_number.encode().as_slice(), i.encode().as_slice()].concat();
+			let optional_decoded_event = get_offchain_index::<Event<T>>(key.encode().as_slice());
+			if let Some(decoded_event) = optional_decoded_event {
+				events.push(decoded_event);
+			}
+			// read and clear the event from offchain storage
+			remove_offchain_index(key.encode().as_slice());
```

Comment:
Once done clear out the counter so no other offchain worker process the same block again

Reply:
Now that I think about it maybe it might make sense to have a block_number level offchain lock instead of msa level lock for 3 reasons

1. From user perspective we only want to check to see if this reverse map is updated up to which block number for all inputs but the way we are storing the block number of storage per msa it's hard to tell that if there are more events after this block number that didn't get indexed yet.
2. since we are removing this storage before processing all the events, there is always that possibility that node might die and there is no trace in offchain DB to continue the process.
3. Having point number 2 in mind we will need to be able to process unprocessed events from previous block numbers also. Since we are not sure that the offchain worker will run on every block number and we might need to resume process for a failed one.

I don't think we might need these at this point but it's some nice to haves.

Reply:
1. Getting the indexed data for a given msa, we get the block_number which is the last updated. Not sure about the overhead
2. Agree a little chance that node crashed before offchain_worker was called, in low transaction world it is instantaneous but i am sure there will be lags. However yes if we can reproduce that else would be worth keeping eye on though we never expect mainnet to stall 
3. Yep we can think of some manual trigger to process missing blocks, but will keep it for laters

Awesome thanks for the amazing review, I will add some unit tests since benchmarks are fairly straightforward